### PR TITLE
Documented: fish.inc.c and blue_fish.inc.c added: defines/fields

### PR DESCRIPTION
**Pull request by MegaMech**
This time for sure.

US Match OK!

### CODE REVIEW

File: src/game/behaviors/fish.inc.c [hunk 1]
```diff
@@ -1,163 +1,233 @@
-// fish.c.inc
+/**
+ * Behaviour and spawner for Fish located in the Secret Aquarium and other levels.
```
Comment (JoshDuMan) [inline src/game/behaviors/fish.inc.c]: @file to designate the file, no need for the blank comment line

File: src/game/behaviors/fish.inc.c [hunk 2]
```diff
@@ -1,163 +1,233 @@
...
+    s32 schoolQuantity;
     s16 model;
-    f32 sp30;
-    struct Animation **sp2C;
-    struct Object *sp28;
+    f32 minDistToMario;
+    struct Animation **fishAnimation;
+    struct Object *fishObject;
     switch (o->oBehParams2ndByte) {
         case 0:
```
Comment (JoshDuMan) [inline src/game/behaviors/fish.inc.c]: This switch case could use defines to possible clarify what each case is.

File: src/game/behaviors/fish.inc.c [hunk 3]
```diff
@@ -1,163 +1,233 @@
...
         o->oAction = 1;
     }
 }
 
-void ActionFish1(void) {
-    if (gCurrLevelNum != LEVEL_SA)
-        if (gMarioObject->oPosY - o->oPosY > 2000.0f)
+// If the current level is not Secret Aquarium and the distance from
+// mario minus o->oPosY is greater than 2000.0f then remove fish.
+void bhv_fish_action_1(void) {
```
Comment (JoshDuMan) [inline src/game/behaviors/fish.inc.c]: This is not a sufficient name. I realize that neither is the old one, but it would be worth taking care of it here. Additionally, this comment should be a block style comment.

Also, bhv_ is reserved for things called in behavior_data. This should just have the formatting fish_act_XXX

File: src/game/behaviors/fish.inc.c [hunk 4]
```diff
@@ -1,163 +1,233 @@
...
+    if (gCurrLevelNum != LEVEL_SA) {
+        if (gMarioObject->oPosY - o->oPosY > 2000.0f) {
             o->oAction = 2;
+        }
+    }
 }
 
-void ActionFish2(void) {
+// Sets action to zero.
+void bhv_fish_action_2(void) {
```
Comment (JoshDuMan) [inline src/game/behaviors/fish.inc.c]: Same as the above action.

File: src/game/behaviors/fish.inc.c [hunk 5]
```diff
@@ -1,163 +1,233 @@
...
         func_8029ED98(0, 1.0f);
+    }
     if (o->oTimer == 0) {
         o->oForwardVel = RandomFloat() * 2 + 3.0f;
-        if (gCurrLevelNum == LEVEL_SA)
+        if (gCurrLevelNum == LEVEL_SA) {
             o->oFishGroupUnkFC = RandomFloat() * 700.0f;
-        else
+        }
+        else {
```
Comment (JoshDuMan) [inline src/game/behaviors/fish.inc.c]: } else {

File: src/game/behaviors/fish.inc.c [hunk 6]
```diff
@@ -1,163 +1,233 @@
...
-    if (o->oTimer < 10)
+// Moves fish forward at a random velocity.
+// In SA move fish group UnkFC at a slower speed.
+void bhv_fish_group_1(void) {
+    f32 fishY = o->oPosY - gMarioObject->oPosY;
+    if (o->oTimer < 10) {
         func_8029ED98(0, 2.0f);
-    else
+    }
+    else {
```
Comment (JoshDuMan) [inline src/game/behaviors/fish.inc.c]: } else {

File: src/game/behaviors/fish.inc.c [hunk 7]
```diff
@@ -1,163 +1,233 @@
...
-            func_802BF59C(2);
-        else
-            func_802BF59C(4);
+        if (fishY < 0.0f) {
+            fishY = 0.0f - fishY;
+        }
+        if (fishY < 500.0f) {
+            regroup_fish(2);
+        }
+        else {
```
Comment (JoshDuMan) [inline src/game/behaviors/fish.inc.c]: } else {

File: src/game/behaviors/fish.inc.c [hunk 8]
```diff
@@ -1,163 +1,233 @@
...
+        o->oPosY = approach_f32_symmetric(o->oPosY, o->oFishGroupUnkF8, speed);
+    }
 }
 
-void ActionFishGroup1(void) {
-    f32 sp1C = o->oPosY - gMarioObject->oPosY;
-    if (o->oTimer < 10)
+// Moves fish forward at a random velocity.
+// In SA move fish group UnkFC at a slower speed.
+void bhv_fish_group_1(void) {
```
Comment (JoshDuMan) [inline src/game/behaviors/fish.inc.c]: See bhv_fish_action_1, just with fish_group_act_XXX instead.

File: src/game/behaviors/fish.inc.c [hunk 9]
```diff
@@ -1,163 +1,233 @@
...
+    // Remove fish if its distance to mario is smaller than his distance to Unk104 + 150.0f
+    if (o->oDistanceToMario < o->oFishGroupUnk104 + 150.0f) {
         o->oAction = 2;
+    }
 }
 
-void ActionFishGroup2(void) {
-    f32 sp1C = o->oPosY - gMarioObject->oPosY;
-    s32 sp18;
+void bhv_fish_group_2(void) {
```
Comment (JoshDuMan) [inline src/game/behaviors/fish.inc.c]: See bhv_fish_action_1, just with fish_group_act_XXX instead.

File: src/game/behaviors/fish.inc.c [hunk 10]
```diff
@@ -1,163 +1,233 @@
...
-        else
-            sp18 = (s32)(1.0 / (o->oDistanceToMario / 600.0));
-        sp18 *= 127;
-        PlaySound2(SOUND_GENERAL_MOVING_WATER);
-    }
-    if (o->oTimer < LEVEL_SA)
+        if (o->oDistanceToMario < 600.0f) {
+            distance = 1;
+        }
+        else {
```
Comment (JoshDuMan) [inline src/game/behaviors/fish.inc.c]: } else {

File: src/game/behaviors/fish.inc.c [hunk 11]
```diff
@@ -1,163 +1,233 @@
...
+            distance = (s32)(1.0 / (o->oDistanceToMario / 600.0));
+        }
+            distance *= 127;
+            PlaySound2(SOUND_GENERAL_MOVING_WATER);
+    }
+    if (o->oTimer < LEVEL_SA) {
         func_8029ED98(0, 4.0f);
-    else
+    }
+    else {
```
Comment (JoshDuMan) [inline src/game/behaviors/fish.inc.c]: } else {

File: src/game/behaviors/fish.inc.c [hunk 12]
```diff
@@ -1,163 +1,233 @@
...
-            func_802BF59C(2);
-        else
-            func_802BF59C(4);
+        if (fishY < 0.0f) {
+            fishY = 0.0f - fishY;
+        }
+        if (fishY < 500.0f) {
+            regroup_fish(2);
+        }
+        else {
```
Comment (JoshDuMan) [inline src/game/behaviors/fish.inc.c]: } else {

File: src/game/behaviors/fish.inc.c [hunk 13]
```diff
@@ -1,163 +1,233 @@
...
     }
-    if (o->oDistanceToMario > o->oFishGroupUnk110 + 500.0f)
+    if (o->oDistanceToMario > o->oFishGroupUnk110 + 500.0f) {
         o->oAction = 1;
+    }
 }
-
-void ActionFishGroup0(void) {
+// Randomly move/animate/scale fish
+void bhv_fish_group_0(void) {
```
Comment (JoshDuMan) [inline src/game/behaviors/fish.inc.c]: See bhv_fish_action_1, just with fish_group_act_XXX instead.

File: src/game/behaviors/fish.inc.c [hunk 14]
```diff
@@ -1,163 +1,233 @@
...
     obj_scale(RandomFloat() * 0.4 + 0.8);
     o->oAction = 1;
 }
 
-void (*sFishGroupActions[])(void) = { ActionFishGroup0, ActionFishGroup1, ActionFishGroup2 };
+// Array of methods
+void (*sFishGroupActions[])(void) = { bhv_fish_group_0, bhv_fish_group_1, bhv_fish_group_2 };
 
+// In SA alter fish pathing if they hit a wall.
 void bhv_fish_group_2_loop(void) // TODO rename
```
Comment (JoshDuMan) [inline src/game/behaviors/fish.inc.c]: Finishing file documentation should handle all TODO's, if thats the goal of this PR.

File: src/game/behaviors/fish.inc.c [hunk 15]
```diff
@@ -1,163 +1,233 @@
...
     UNUSED s32 unused[4];
     obj_scale(1.0f);
     o->oFishGroupUnkF4 = find_water_level(o->oPosX, o->oPosZ);
-    if (gCurrLevelNum == LEVEL_SA)
+    if (gCurrLevelNum == LEVEL_SA) {
         o->oFishGroupUnkF4 = 0.0f;
+    }
     o->oWallHitboxRadius = 30.0f;
     obj_resolve_wall_collisions();
+    // Cleanup fish if they are nil
```
Comment (JoshDuMan) [inline src/game/behaviors/fish.inc.c]: This is slightly misleading, LEVEL_UNKNOWN_32 is not a used level so this is technically dead code. Perhaps a note on that?

Comment (MegaMech) [reply src/game/behaviors/fish.inc.c]: Perhaps it's checking the water level of all levels?

File: src/game/behaviors/fish.inc.c [hunk 16]
```diff
@@ -1,163 +1,286 @@
-// fish.c.inc
+/**
+ * @file fish.inc.c
+ * Implements behaviour and spawning for fish located in the Secret Aquarium and other levels.
+ *
```
Comment (JoshDuMan) [inline src/game/behaviors/fish.inc.c]: Unneeded blank comment line.

File: src/game/behaviors/fish.inc.c [hunk 17]
```diff
@@ -1,163 +1,286 @@
...
 }
 
-void ActionFish1(void) {
-    if (gCurrLevelNum != LEVEL_SA)
-        if (gMarioObject->oPosY - o->oPosY > 2000.0f)
+/**
+ * If the current level is not Secret Aquarium and the distance from
+ * mario minus o->oPosY is greater than 2000.0f then set action to two.
+ */
+void fish_action_chkDistance(void) {
```
Comment (JoshDuMan) [inline src/game/behaviors/fish.inc.c]: snake_case

i.e. fish_action_chk_distance

However, 2 things-
1) write out check
2) make either _action_ or _act_ consistent. _act_ is preferable imo

File: src/game/behaviors/fish.inc.c [hunk 18]
```diff
@@ -1,163 +1,286 @@
...
-        if (500.0f < absf(o->oPosY - o->oFishGroupUnkF8))
-            a0 = 10;
-        o->oPosY = approach_f32_symmetric(o->oPosY, o->oFishGroupUnkF8, a0);
-    } else if (sp1C - 100.0f - o->oFishGroupUnk10C < o->oPosY
-               && o->oPosY < sp1C + 1000.0f + o->oFishGroupUnk10C)
-        o->oPosY = approach_f32_symmetric(o->oPosY, o->oFishGroupUnkF8, a0);
+        if (500.0f < absf(o->oPosY - o->oFishGroupUnkF8)) {
+            speed = 10;
+        }
+            //Move fish group heighth if level is SA at incremented speed.
```
Comment (JoshDuMan) [inline src/game/behaviors/fish.inc.c]: heighth

File: src/game/behaviors/fish.inc.c [hunk 19]
```diff
@@ -1,163 +1,286 @@
...
+void fish_action_chkDistance(void) {
+    if (gCurrLevelNum != LEVEL_SA) {
+        if (gMarioObject->oPosY - o->oPosY > 2000.0f) {
             o->oAction = 2;
+        }
+    }
 }
 
-void ActionFish2(void) {
+/*
```
Comment (JoshDuMan) [inline src/game/behaviors/fish.inc.c]: Block comments should start with /**

File: src/game/behaviors/fish.inc.c [hunk 20]
```diff
@@ -1,163 +1,286 @@
...
+ * Sets fish->action to zero.
+ */
+void fish_action_set(void) {
     o->oAction = 0;
 }
 
-void (*sFishActions[])(void) = { ActionFish0, ActionFish1, ActionFish2 };
+/*
```
Comment (JoshDuMan) [inline src/game/behaviors/fish.inc.c]: /**

File: src/game/behaviors/fish.inc.c [hunk 21]
```diff
@@ -1,163 +1,286 @@
...
+ * An array of action methods
+ */
+void (*sFishActions[])(void) = { fish_act_spawn, fish_action_chkDistance, fish_action_set };
 
+/*
```
Comment (JoshDuMan) [inline src/game/behaviors/fish.inc.c]: /**

File: src/game/behaviors/fish.inc.c [hunk 22]
```diff
@@ -1,163 +1,286 @@
...
+ * Calls an array of actions based on the value of o->oAction
+ */
 void bhv_fish_loop(void) {
     obj_call_action_function(sFishActions);
 }
 
-void func_802BF59C(s32 a0) {
-    f32 sp1C = o->parentObj->oPosY;
+/*
```
Comment (JoshDuMan) [inline src/game/behaviors/fish.inc.c]: /**

File: src/game/behaviors/fish.inc.c [hunk 23]
```diff
@@ -1,163 +1,286 @@
...
+        if (o->oDistanceToMario < 600.0f) {
+            distance = 1;
+        } else {
+            distance = (s32)(1.0 / (o->oDistanceToMario / 600.0));
+        }
+        distance *= 127;
         PlaySound2(SOUND_GENERAL_MOVING_WATER);
     }
-    if (o->oTimer < LEVEL_SA)
+    // Enable fish animation. Parameters differ if oTimer < SA.
```
Comment (JoshDuMan) [inline src/game/behaviors/fish.inc.c]: This is technically a repo typo. LEVEL_SA should just be 20 here, it is fixed privately.

File: include/object_constants.h [hunk 24]
```diff
@@ -197,6 +197,19 @@
...
+    #define FISH_INIT 0
+    #define FISH_ACTIVE 1
+    #define FISH_RESPAWN 2
+
+/* Blue_Fish */
+    /* oAction */
+    #define BLUE_FISH_SPAWN 0
+    #define BLUE_FISH_RANDOM_YAW 1
+    #define BLUE_FISH_RANDOM_PITCH 2
+    #define BLUE_FISH_RANDOM_YAW_2 3
```
Comment (nim-ka) [inline include/object_constants.h]: I don't like having RANDOM_YAW and RANDOM_YAW_2, it's confusing and not descriptive. What's the difference between these?

File: include/object_fields.h [hunk 25]
```diff
@@ -5,7 +5,7 @@
 /**
  * The array [0x88, 0x1C8) in struct Object consists of fields that can vary by
  * object type. These macros provide access to these fields.
- */
+ */
```
Comment (nim-ka) [inline include/object_fields.h]: Trailing space

File: include/object_fields.h [hunk 26]
```diff
@@ -483,19 +483,19 @@
...
-#define /*0x10C*/ oFishGroupUnk10C OBJECT_FIELD_F32(0x21)
-#define /*0x110*/ oFishGroupUnk110 OBJECT_FIELD_F32(0x22)
+#define /*0x0F4*/ oFishWaterLevel      OBJECT_FIELD_F32(0x1B)
+#define /*0x0F8*/ oFishPosY            OBJECT_FIELD_F32(0x1C)
+#define /*0x0FC*/ oFishRandomOffset    OBJECT_FIELD_F32(0x1D)
+#define /*0x100*/ oFishRandomSpeed     OBJECT_FIELD_S32(0x1E)
+#define /*0x104*/ oFishRandomDistance  OBJECT_FIELD_F32(0x1F)
+#define /*0x108*/ oFishRandomVel       OBJECT_FIELD_F32(0x20)
+#define /*0x10C*/ oFishRandomYDistance OBJECT_FIELD_F32(0x21)
+#define /*0x110*/ oFishRandomDistance2 OBJECT_FIELD_F32(0x22)
```
Comment (nim-ka) [inline include/object_fields.h]: Again, I don't like RandomDistance and RandomDistance2. Also you have RandomYDistance which is kinda confusing when reading code at a glance, since there are 3 fields which are named almost the same thing

File: src/game/behaviors/blue_fish.inc.c [hunk 27]
```diff
@@ -1,76 +1,131 @@
...
             break;
     }
+    
+    // Calculates Y velocity and calls physics engine.
     o->oVelY = -sins(o->oFaceAnglePitch) * o->oForwardVel;
     obj_move_using_fvel_and_gravity();
-    if (o->parentObj->oAction == 2)
+    
+    // Deletes object if the parent has oAction set to two.
+    if (o->parentObj->oAction == 2) {
```
Comment (nim-ka) [inline src/game/behaviors/blue_fish.inc.c]: Label oAction here, isn't the parent object the fish group which is in the same file? GitHub reviewing system doesn't let me see the whole file lol

File: include/object_constants.h (hunk 24 above)
Comment (nim-ka) [reply include/object_constants.h]: Also you forgot _ACT_, i.e. `BLUE_FISH_ACT_RANDOM_YAW`

File: src/game/behaviors/blue_fish.inc.c [hunk 28]
```diff
@@ -1,76 +1,131 @@
...
+            }
             break;
-        case 2:
+            
+        // Animates and adjusts pitch velocity.
+        case BLUE_FISH_RANDOM_PITCH:
             func_8029ED98(0, 1.0f);
-            if (o->oTimer >= o->oBlueFishUnkF8 + 60)
+            
+            // Progresses oActioon after some time.
```
Comment (nim-ka) [inline src/game/behaviors/blue_fish.inc.c]: Typo. Also I'll use this comment to say that "Progresses oAction" is a useless comment, since everyone knows what ++ means. It would be much more helpful to say what action is being progressed to, i.e. `// Switches to BLUE_FISH_ACT_RANDOM_PITCH`

File: src/game/behaviors/fish.inc.c [hunk 29]
```diff
@@ -1,163 +1,299 @@
...
     s16 model;
-    f32 sp30;
-    struct Animation **sp2C;
-    struct Object *sp28;
+    f32 minDistToMario;
+    struct Animation **fishAnimation;
+    struct Object *fishObject;
     switch (o->oBehParams2ndByte) {
+        
+        // Blue fish with a quanitiy of twenty.
```
Comment (nim-ka) [inline src/game/behaviors/fish.inc.c]: Typo. I'm probably not catching all typos so it might be worth running a spellcheck

File: src/game/behaviors/fish.inc.c [hunk 30]
```diff
@@ -1,163 +1,299 @@
...
+            fishAnimation = blue_fish_seg3_anims_0301C2B0;
             break;
+            
+        // Cyan fish with a quanitiy of twenty.
         case 2:
-            model = MODEL_CYAN_FISH;            sp38 = 20;            sp30 = 1500.0f;            sp2C = cyan_fish_seg6_anims_0600E264;
+            model = MODEL_CYAN_FISH;
+            schoolQuantity = 20;
+            minDistToMario = 1500.0f;
+            fishAnimation = cyan_fish_seg6_anims_0600E264;
```
Comment (nim-ka) [inline src/game/behaviors/fish.inc.c]: NO NO NO you have to undo this. These lines need to be on the same line for PAL version to match. Same for all the other cases.

File: src/game/behaviors/fish.inc.c [hunk 31]
```diff
@@ -1,163 +1,299 @@
...
+        if (gMarioObject->oPosY - o->oPosY > 2000.0f) {
+            o->oAction = FISH_RESPAWN;
+        }
+    }
 }
 
-void ActionFish2(void) {
-    o->oAction = 0;
+/**
+ * set oAction flag to FISH_INIT.
```
Comment (nim-ka) [inline src/game/behaviors/fish.inc.c]: Capitalization. Also, your action macros should match your action function names. Ex. `FISH_ACT_RESPAWN` is a fine name but where's the `fish_act_respawn` function?

File: src/game/behaviors/fish.inc.c [hunk 32]
```diff
@@ -1,163 +1,299 @@
...
+ */
+void fish_act_set(void) {
```
Comment (nim-ka) [inline src/game/behaviors/fish.inc.c]: I don't like the names `fish_act_check_distance` or `fish_act_set`. I think the names you used in the action macros are much better.

File: src/game/behaviors/fish.inc.c [hunk 33]
```diff
@@ -1,163 +1,299 @@
...
-    } else if (sp1C - 100.0f - o->oFishGroupUnk10C < o->oPosY
-               && o->oPosY < sp1C + 1000.0f + o->oFishGroupUnk10C)
-        o->oPosY = approach_f32_symmetric(o->oPosY, o->oFishGroupUnkF8, a0);
+/**
+ * Move object up or down towards oFishPosY by the speed variable
+ * In Secret Aquarium, the speed variable is always ten.
+ */
+void fish_regroup(s32 speed) {
+    // Set pPosY to the parentY.
+    f32 pPosY = o->parentObj->oPosY;
```
Comment (nim-ka) [inline src/game/behaviors/fish.inc.c]: minor nitpick but `parentY` is more descriptive than `pPosY`

File: src/game/behaviors/fish.inc.c [hunk 34]
```diff
@@ -1,163 +1,299 @@
...
+    
+    // If level is SA and fish height minus oPosY is higher than 500 then speed = 10.
```
Comment (nim-ka) [inline src/game/behaviors/fish.inc.c]: You're describing what the code does, but not *why* it does it. This is also an issue in lots of other places in the files, but it's more of an abstract issue so it's hard to show an outright example.

File: src/game/behaviors/fish.inc.c [hunk 35]
```diff
@@ -1,163 +1,299 @@
...
+            speed = 10;
+        }
+        // Increase fish Y coord at the incremented speed.
+        o->oPosY = approach_f32_symmetric(o->oPosY, o->oFishPosY, speed);
+     /**
+     * For other levels, if oFishRandomYDistance - pPosy - 100.0f is lower than 
+     * o->oPosY - 100 and * o->PosY is lower than the pPosY + 1000.0f +
+     * oFishRandomYDistance then have the fish 
+     * symmetrically approach the target oFishPosY.
+     */
```
Comment (nim-ka) [inline src/game/behaviors/fish.inc.c]: should be
```
/**
 *
 *
 *
 */
```
instead of
```
/**
*
*
*
*/
```
also again you're not explaining *why* the code does that

File: src/game/behaviors/fish.inc.c [hunk 36]
```diff
@@ -1,163 +1,299 @@
...
+     * Assigns oForwardVel, oFishRandomOffset, & oFishRandomDistance to a random floats.
+     */
     if (o->oTimer == 0) {
         o->oForwardVel = RandomFloat() * 2 + 3.0f;
-        if (gCurrLevelNum == LEVEL_SA)
-            o->oFishGroupUnkFC = RandomFloat() * 700.0f;
-        else
-            o->oFishGroupUnkFC = RandomFloat() * 100.0f;
-        o->oFishGroupUnk104 = RandomFloat() * 500 + 200.0f;
+        if (gCurrLevelNum == 20) {
```
Comment (nim-ka) [inline src/game/behaviors/fish.inc.c]: Why did you remove the LEVEL_SA macro?

File: src/game/behaviors/fish.inc.c [hunk 37]
```diff
@@ -1,163 +1,299 @@
...
+            o->oFishRandomOffset = RandomFloat() * 700.0f;
+        } else {
+            o->oFishRandomOffset = RandomFloat() * 100.0f;
+        }
+        o->oFishRandomDistance = RandomFloat() * 500 + 200.0f;
     }
-    o->oFishGroupUnkF8 = gMarioObject->oPosY + o->oFishGroupUnkFC;
+    
+    // Rotate fish towards mario at a rate of 5.625 degrees (0x400)
```
Comment (nim-ka) [inline src/game/behaviors/fish.inc.c]: Capitalize Mario, also only use Hexadecimal Angle Units (HAU) here since the value was pretty clearly not originally in degrees (5.625?). In general only reference degrees if the value was clearly converted from degrees to HAU

File: src/game/behaviors/fish.inc.c [hunk 38]
```diff
@@ -1,163 +1,299 @@
...
+            distance = 1;
+        } else {
+            distance = (s32)(1.0 / (o->oDistanceToMario / 600.0));
+        }
+        distance *= 127;
         PlaySound2(SOUND_GENERAL_MOVING_WATER);
     }
-    if (o->oTimer < LEVEL_SA)
+    // Enable fish animation. Parameters differ if oTimer < SA.
+    if (o->oTimer < 20) {
```
Comment (nim-ka) [inline src/game/behaviors/fish.inc.c]: Oh is this why you changed LEVEL_SA earlier? You were right to change it here, but before, it was a proper reference to the level

File: src/game/behaviors/fish.inc.c [hunk 39]
```diff
@@ -1,163 +1,299 @@
...
 
-void (*sFishGroupActions[])(void) = { ActionFishGroup0, ActionFishGroup1, ActionFishGroup2 };
+/**
+ * Array of methods
+ */
+void (*sFishGroupActions[])(void) = { 
+    fish_group_act_animate, 
+    fish_group_act_rotation, 
+    fish_group_act_move
+};
```
Comment (nim-ka) [inline src/game/behaviors/fish.inc.c]: Why is sFishGroupActions on 4 lines but sFishActions on 1? Minor nitpick, but consistency would be nice

File: src/game/behaviors/fish.inc.c [hunk 40]
```diff
@@ -1,163 +1,299 @@
...
 
-void bhv_fish_group_2_loop(void) // TODO rename
+/**
+ * Main loop for fish
+ */
+void bhv_fish_group_2_loop(void)
```
Comment (nim-ka) [inline src/game/behaviors/fish.inc.c]: Wait. wait. wait.

`bhv_fish_group_2_loop` is the main loop for fish, instead of `bhv_fish_loop`???
What is the _2 doing in that name?
If it's not the actual fish behavior, why does the fish *group* use object fields with prefix oFish, and have action macros like `FISH_RESPAWN` instead of `FISH_GROUP_RESPAWN`?
This is very confusing.

File: src/game/behaviors/fish.inc.c [hunk 41]
```diff
@@ -1,163 +1,299 @@
...
 
-void (*sFishActions[])(void) = { ActionFish0, ActionFish1, ActionFish2 };
+/**
+ * An array of action methods
+ */
+void (*sFishActions[])(void) = { fish_act_spawn, fish_act_check_distance, fish_act_set };
 
+/**
+ * Calls an array of actions based on the value of o->oAction
+ */
```
Comment (nim-ka) [inline src/game/behaviors/fish.inc.c]: Unnecessary comment.

File: src/game/behaviors/fish.inc.c (hunk 40 above)
Comment (nim-ka) [reply src/game/behaviors/fish.inc.c]: I think the *reality* might be something similar to birds, where there's 1 "leader" bird that's loaded from the start of the level and spawns in all the other birds, but otherwise acts the same as a regular bird. Is this a correct guess?

File: include/object_constants.h (hunk 24 above)
Comment (MegaMech) [reply include/object_constants.h]: Random yaw and random yaw 2 are near identical

File: src/game/behaviors/blue_fish.inc.c (hunk 27 above)
Comment (MegaMech) [reply src/game/behaviors/blue_fish.inc.c]: 3 dots -> view file

I do not know what the parentObject is.

File: src/game/behaviors/fish.inc.c (hunk 38 above)
Comment (MegaMech) [reply src/game/behaviors/fish.inc.c]: Did I misunderstand that all "LEVEL_SA" should just be 20?

File: src/game/behaviors/fish.inc.c (hunk 40 above)
Comment (MegaMech) [reply src/game/behaviors/fish.inc.c]: I will consider this next time I reading through the file.
I have not changed the names of BHV_ methods because they are predetermined in another file. Of course I can change that one too. Just for now I haven't. Should I change it?

File: src/game/behaviors/blue_fish.inc.c (hunk 27 above)
Comment (nim-ka) [reply src/game/behaviors/blue_fish.inc.c]: then investigate!

File: src/game/behaviors/fish.inc.c (hunk 38 above)
Comment (nim-ka) [reply src/game/behaviors/fish.inc.c]: yes, he was only talking about that one, since it's obviously not a level reference.

Just think about it: which of these makes sense?
`myObject.position == POSITION_VECTOR_1`
`myObject.position == 20`
`myObject.speed == POSITION_VECTOR_1`
`myObject.speed == 20`
replace position with level and speed with timer and it's the same situation

File: src/game/behaviors/fish.inc.c (hunk 40 above)
Comment (nim-ka) [reply src/game/behaviors/fish.inc.c]: Yes of course! Otherwise your documentation and the function's names conflict with each other.

File: include/object_constants.h (hunk 24 above)
Comment (Synray) [reply include/object_constants.h]: maybe (BLUE_FISH_ACT_)`DIVE`, `TURN`, `ASCEND`, and `TURN_BACK` are more descriptive.
unk100 is always +-0x800, so the fish turns 0x8000 (180 degrees) during 1, and again during 3. rng just decides left or right

File: src/game/behaviors/blue_fish.inc.c [hunk 42]
```diff
@@ -1,76 +1,131 @@
...
+/**
+ * Spawns fifteen fish if mario resides in room fifteen or seven.
+ * It moves at random within 200.0f
+ */
+void bhv_blue_fish_spawn_loop(void) {
     struct Object *fish;
     s32 i;
     switch (o->oAction) {
-        case 0:
+        case BLUE_FISH_ACT_DIVE:
```
Comment (Synray) [inline src/game/behaviors/blue_fish.inc.c]: this spawn loop is used by a different object, so it needs to use different names. (I'd change the function back to fish_tank or something as well.)

File: src/game/behaviors/fish.inc.c (hunk 30 above)
Comment (MegaMech) [reply src/game/behaviors/fish.inc.c]: But line limit of 104. (Didn't see this yesterday, will fix)

File: src/game/behaviors/blue_fish.inc.c (hunk 42 above)
Comment (MegaMech) [reply src/game/behaviors/blue_fish.inc.c]: bhv_tank isn't consistent with the other method names of blue_fish_ and tank doesn't really describe what the function is.
Unless the clarification of tank helps to differentiate with other fish methods?

Comment (Synray) [reply src/game/behaviors/blue_fish.inc.c]: It's fine if it's not consistent with blue_fish_, because it's a different object. Sorry if I was unclear, this is the main loop for `bhvTankFishGroup`, which is what spawns the fish in the JRB room. To be consistent with the rest of the code base, this function should be named `bhv_tank_fish_group_loop`. (or rename bhvTankFishGroup to something better like `bhvBlueFishTank`/`Spawn`)

But because this is the TankFishGroup's (or whatever you want to rename it)'s main loop, `BLUE_FISH_ACT_DIVE` and etc. shouldn't be reused in this function.

**Comment by MegaMech** [comment]
Thanks for your help! Closed for now.
